File: monitoring/synthetic-monitoring.js
'use strict';

const opentelemetry = require('@opentelemetry/api');
const { MeterProvider } = require('@opentelemetry/sdk-metrics');
const { OTLPMetricExporter } = require('@opentelemetry/exporter-metrics-otlp-grpc');
const axios = require('axios');

class SyntheticMonitor {
    constructor(serviceName = 'synthetic-monitor') {
        this.serviceName = serviceName;
        this.setupTelemetry();
    }

    setupTelemetry() {
        // Configure metrics
        const meterProvider = new MeterProvider();
        const metricExporter = new OTLPMetricExporter({
            url: 'http://otel-collector:4317'
        });

        meterProvider.addMetricReader(metricExporter);
        opentelemetry.metrics.setGlobalMeterProvider(meterProvider);

        const meter = opentelemetry.metrics.getMeter(this.serviceName);

        // Create metrics
        this.requestDuration = meter.createHistogram('synthetic_request_duration', {
            description: 'Duration of synthetic requests',
            unit: 'ms'
        });

        this.errorCounter = meter.createCounter('synthetic_request_errors', {
            description: 'Number of synthetic request errors'
        });
    }

    async checkApiEndpoint(url, method = 'GET', expectedStatus = 200) {
        const startTime = Date.now();
        try {
            const response = await axios({
                method,
                url,
                validateStatus: null
            });

            const duration = Date.now() - startTime;
            this.requestDuration.record(duration, {
                endpoint: url,
                method
            });

            if (response.status !== expectedStatus) {
                this.errorCounter.add(1, {
                    endpoint: url,
                    error: 'status_mismatch'
                });
                return false;
            }
            return true;
        } catch (error) {
            this.errorCounter.add(1, {
                endpoint: url,
                error: error.message
            });
            return false;
        }
    }

    async generateAiLoad(modelEndpoint, batchSize = 1) {
        // Create random input data
        const inputData = Array.from({ length: batchSize }, () =>
            Array.from({ length: 3 }, () =>
                Array.from({ length: 224 }, () =>
                    Array.from({ length: 224 }, () =>
                        Math.random() * 2 - 1
                    )
                )
            )
        );

        const startTime = Date.now();
        try {
            const response = await axios.post(modelEndpoint, {
                inputs: inputData
            });

            const duration = Date.now() - startTime;
            this.requestDuration.record(duration, {
                endpoint: modelEndpoint,
                type: 'inference'
            });

            if (response.status !== 200) {
                this.errorCounter.add(1, {
                    endpoint: modelEndpoint,
                    error: 'inference_failed'
                });
            }
        } catch (error) {
            this.errorCounter.add(1, {
                endpoint: modelEndpoint,
                error: error.message
            });
        }
    }

    async triggerHomeAutomation(hassUrl, token) {
        const headers = {
            Authorization: `Bearer ${token}`,
            'Content-Type': 'application/json'
        };

        const scenarios = [
            { service: 'light/turn_on', entity_id: 'light.living_room' },
            { service: 'switch/toggle', entity_id: 'switch.office_fan' },
            { service: 'scene/turn_on', entity_id: 'scene.evening_mode' }
        ];

        for (const scenario of scenarios) {
            const startTime = Date.now();
            try {
                const response = await axios.post(
                    `${hassUrl}/api/services/${scenario.service}`,
                    { entity_id: scenario.entity_id },
                    { headers }
                );

                const duration = Date.now() - startTime;
                this.requestDuration.record(duration, {
                    service: scenario.service,
                    entity: scenario.entity_id
                });

                if (response.status !== 200) {
                    this.errorCounter.add(1, {
                        service: scenario.service,
                        error: 'automation_failed'
                    });
                }
            } catch (error) {
                this.errorCounter.add(1, {
                    service: scenario.service,
                    error: error.message
                });
            }
        }
    }
}

async function main() {
    const monitor = new SyntheticMonitor();

    async function runMonitoring() {
        try {
            // Check API endpoints
            await monitor.checkApiEndpoint('http://api-gateway/health');
            await monitor.checkApiEndpoint('http://rabbitmq:15672/api/health');

            // Generate AI model load
            await monitor.generateAiLoad('http://triton:8000/v2/models/resnet50/infer');

            // Trigger home automation
            await monitor.triggerHomeAutomation(
                'http://homeassistant:8123',
                'YOUR_LONG_LIVED_ACCESS_TOKEN'
            );
        } catch (error) {
            console.error('Error in monitoring loop:', error);
        }
    }

    // Run monitoring every minute
    setInterval(runMonitoring, 60000);
    runMonitoring(); // Run immediately on start
}

if (require.main === module) {
    main().catch(console.error);
}

module.exports = SyntheticMonitor;

File: monitoring/synthetic-monitoring.py
import asyncio
import json
import logging
import random
import time
from typing import Dict, List
from datetime import datetime

import aiohttp
import numpy as np
from opentelemetry import trace, metrics
from opentelemetry.trace import Status, StatusCode
from opentelemetry.metrics import Histogram, Counter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader

# Initialize logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class SyntheticMonitor:
    def __init__(self, service_name: str = "synthetic-monitor"):
        self.service_name = service_name
        self._setup_telemetry()
        self.session = None

    def _setup_telemetry(self):
        """Set up OpenTelemetry instrumentation."""
        # Configure metrics
        reader = PeriodicExportingMetricReader(
            OTLPMetricExporter(endpoint="http://otel-collector:4317")
        )
        provider = MeterProvider(metric_readers=[reader])
        metrics.set_meter_provider(provider)
        self.meter = metrics.get_meter(self.service_name)

        # Create metrics
        self.request_duration = self.meter.create_histogram(
            name="synthetic_request_duration",
            description="Duration of synthetic requests",
            unit="ms"
        )
        self.error_counter = self.meter.create_counter(
            name="synthetic_request_errors",
            description="Number of synthetic request errors"
        )

    async def setup(self):
        """Set up HTTP session."""
        self.session = aiohttp.ClientSession()

    async def cleanup(self):
        """Clean up resources."""
        if self.session:
            await self.session.close()

    async def check_api_endpoint(self, url: str, method: str = "GET", expected_status: int = 200) -> bool:
        """Check an API endpoint."""
        start_time = time.time()
        try:
            async with self.session.request(method, url) as response:
                duration = (time.time() - start_time) * 1000
                self.request_duration.record(
                    duration,
                    {"endpoint": url, "method": method}
                )

                if response.status != expected_status:
                    self.error_counter.add(1, {"endpoint": url, "error": "status_mismatch"})
                    return False
                return True
        except Exception as e:
            self.error_counter.add(1, {"endpoint": url, "error": str(e)})
            return False

    async def generate_ai_load(self, model_endpoint: str, batch_size: int = 1):
        """Generate synthetic load for AI models."""
        # Create random input data
        input_data = np.random.randn(batch_size, 3, 224, 224).tolist()

        start_time = time.time()
        try:
            async with self.session.post(
                model_endpoint,
                json={"inputs": input_data}
            ) as response:
                duration = (time.time() - start_time) * 1000
                self.request_duration.record(
                    duration,
                    {"endpoint": model_endpoint, "type": "inference"}
                )

                if response.status != 200:
                    self.error_counter.add(
                        1,
                        {"endpoint": model_endpoint, "error": "inference_failed"}
                    )
        except Exception as e:
            self.error_counter.add(
                1,
                {"endpoint": model_endpoint, "error": str(e)}
            )

    async def trigger_home_automation(self, hass_url: str, token: str):
        """Trigger Home Assistant automations."""
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json"
        }

        # List of test scenarios
        scenarios = [
            {"service": "light/turn_on", "entity_id": "light.living_room"},
            {"service": "switch/toggle", "entity_id": "switch.office_fan"},
            {"service": "scene/turn_on", "entity_id": "scene.evening_mode"}
        ]

        for scenario in scenarios:
            start_time = time.time()
            try:
                async with self.session.post(
                    f"{hass_url}/api/services/{scenario['service']}",
                    headers=headers,
                    json={"entity_id": scenario["entity_id"]}
                ) as response:
                    duration = (time.time() - start_time) * 1000
                    self.request_duration.record(
                        duration,
                        {
                            "service": scenario["service"],
                            "entity": scenario["entity_id"]
                        }
                    )

                    if response.status != 200:
                        self.error_counter.add(
                            1,
                            {
                                "service": scenario["service"],
                                "error": "automation_failed"
                            }
                        )
            except Exception as e:
                self.error_counter.add(
                    1,
                    {
                        "service": scenario["service"],
                        "error": str(e)
                    }
                )

async def main():
    # Initialize monitor
    monitor = SyntheticMonitor()
    await monitor.setup()

    try:
        while True:
            # Check API endpoints
            await monitor.check_api_endpoint("http://api-gateway/health")
            await monitor.check_api_endpoint("http://rabbitmq:15672/api/health")

            # Generate AI model load
            await monitor.generate_ai_load("http://triton:8000/v2/models/resnet50/infer")

            # Trigger home automation
            await monitor.trigger_home_automation(
                "http://homeassistant:8123",
                "YOUR_LONG_LIVED_ACCESS_TOKEN"
            )

            # Wait before next iteration
            await asyncio.sleep(60)  # Run every minute

    except KeyboardInterrupt:
        logger.info("Shutting down synthetic monitoring")
    finally:
        await monitor.cleanup()

if __name__ == "__main__":
    asyncio.run(main())
